File: src/main/resources/templates/javascript/solution/src/policy.js
import BubbleSort from './bubblesort.js';
import MergeSort from './mergesort.js';

const DATES_LENGTH_THRESHOLD = 10;

export default class Policy {
    /** @type {Context} */
    #context;

    /**
     * @param context {Context}
     */
    constructor(context) {
        this.#context = context;
    }

    /**
     * Chooses a strategy depending on the number of date objects.
     */
    configure() {
        if (this.#context.dates.length > DATES_LENGTH_THRESHOLD) {
            this.#context.sortAlgorithm = new MergeSort();
        } else {
            this.#context.sortAlgorithm = new BubbleSort();
        }
    }

    get context() {
        return this.#context;
    }

    set context(context) {
        this.#context = context;
    }
}

File: src/main/resources/templates/ruby/solution/src/policy.rb
# frozen_string_literal: true

require_relative "bubble_sort"
require_relative "merge_sort"

class Policy
  attr_accessor :context

  DATES_SIZE_THRESHOLD = 10

  def initialize(context)
    @context = context
  end

  # Chooses a strategy depending on the number of date objects.
  def configure
    sort_algorithm = if @context.dates.length > DATES_SIZE_THRESHOLD
                       MergeSort.new
                     else
                       BubbleSort.new
                     end
    @context.sort_algorithm = sort_algorithm
  end
end
